### PR TITLE
Add ignition-math

### PR DESCRIPTION
**Pull request by joxoby**
Specify library name and version:  **ignition-math/6.6.0**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/ignition-math/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+class IgnitionMathConan(ConanFile):
+    name = "ignition-math"
+    license = "Apache-2.0"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://ignitionrobotics.org/libs/math"
+    description = " Math classes and functions for robot applications"
+    topics = ("ignition", "math", "robotics", "gazebo")
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"shared": [True, False]}
+    default_options = {"shared": False}
```
Comment (prince-chrismc) [inline recipes/ignition-math/all/conanfile.py:15]: I would have expected an fpic option :thinking:

File: recipes/ignition-math/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    generators = "cmake", "cmake_find_package_multi"
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        version_major = self.version.split('.')[0]
+        os.rename(f"ign-math-ignition-math{version_major}_{self.version}", self._source_subfolder)
```
Comment (prince-chrismc) [inline recipes/ignition-math/all/conanfile.py:25]: this format is not support in the minimum python version CCI supports, we have to use `"".format()`

You can also use the https://docs.conan.io/en/latest/reference/tools.html#tools-version to get the major version

File: recipes/ignition-math/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    def requirements(self):
+        self.requires("eigen/3.3.7")
+
+    def _configure_cmake(self):
+        cmake = CMake(self)
```
Comment (prince-chrismc) [inline recipes/ignition-math/all/conanfile.py:31]: it's usually worth while caching the cmake object to help speed up builds

File: recipes/ignition-math/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        tools.rmdir(os.path.join(self.package_folder, "share"))
+        tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+
+    def package_info(self):
+        version_major = self.version.split('.')[0]
+        self.cpp_info.libs = tools.collect_libs(self)
+        self.cpp_info.names["cmake_find_package"] = f"ignition-math{version_major}"
+        self.cpp_info.names["cmake_find_package_multi"] = f"ignition-math{version_major}"
+        self.cpp_info.includedirs = [f"include/ignition/math{version_major}"]
```
Comment (prince-chrismc) [inline recipes/ignition-math/all/conanfile.py:54]: please use `"".format()`

File: recipes/ignition-math/all/test_package/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,13 @@
+cmake_minimum_required(VERSION 3.1)
```
Comment (prince-chrismc) [inline recipes/ignition-math/all/test_package/CMakeLists.txt:1]: I believe CXX_STANDARD 17 was introduced in 3.8

**Comment by prince-chrismc** [comment]
Since the test_package requires c++17 support, I assume the main project does as well, in that case you will need to set the "minimum compiler support" this is an example that easy to copy-paste https://github.com/conan-io/conan-center-index/blob/2b092c355fdb3d96a368aa3d84f779f54c3b6af6/recipes/tomlplusplus/all/conanfile.py#L27-L51

File: recipes/ignition-math/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        os.rename(
+            "ign-math-ignition-math{}_{}".format(version_major, self.version),
+            self._source_subfolder,
+        )
+
+    def requirements(self):
+        self.requires("eigen/3.3.7")
+
+    def _configure_cmake(self):
+        self.cmake = CMake(self)
```
Comment (prince-chrismc) [inline recipes/ignition-math/all/conanfile.py:73]: You can use this as an example for how to cache the cmake object (to be more consistent with other recipes) https://github.com/conan-io/conan-center-index/pull/3221/files#diff-510d6ea9c94f2cb339e72a289b88ee5a25e0258b7ddd160973a5e4476ef18542R40

File: recipes/ignition-math/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
```
Comment (prince-chrismc) [inline recipes/ignition-math/all/conanfile.py:41]:
```suggestion
    def configure(self):
        if self.options.shared:
            del self.options.fPIC
```

File: recipes/ignition-math/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    name = "ignition-math"
+    license = "Apache-2.0"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://ignitionrobotics.org/libs/math"
+    description = " Math classes and functions for robot applications"
+    topics = ("ignition", "math", "robotics", "gazebo")
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"shared": [True, False], "fPIC": [True, False]}
+    default_options = {"shared": False, "fPIC": True}
+    generators = "cmake", "cmake_find_package_multi"
```
Comment (prince-chrismc) [inline recipes/ignition-math/all/conanfile.py:18]:
```suggestion
    generators = "cmake_find_package_multi"
```

**Comment by joxoby** [comment]
This PR reached the point in which I'm running out of ammunition. I was able to reproduce the error during the run of `test_package`, which happens (I think) during the dynamic loading of `ignition-math6`. The error (running gdb on a clang5 docker container):
```
Program received signal SIGSEGV, Segmentation fault.
__memmove_avx_unaligned_erms () at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:253
253	../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S: No such file or directory.
```
I'm not sure how to keep debugging from here.

**Comment by joxoby** [comment]
After a lot of pain, I was able to solve the seg fault error that was happening. The problem was the static initialization of a `std::regex` variable: https://github.com/ignitionrobotics/ign-math/blob/51eb0640a0c7454c36fef1a3007d3db37fe9165f/include/ignition/math/Helpers.hh#L873, which I solved by creating a raw  static`char*` and using that to initialize a local static (static inside the function) `std::regex` variable: https://github.com/conan-io/conan-center-index/pull/3215/files#diff-e2f831892c4567a7388d8d62242576015a59f008af8dd34c946996d15f6991e9

There is now another problem that is puzzling me, which is the failing build in, for example, this context:
**Linux x86_64, Release, clang 5.0, libc++ . Options: ignition-math:shared-False**
with an error trying to build the test_package:
```
[ 50%] Building CXX object CMakeFiles/test_package.dir/test_package.cpp.o
[100%] Linking CXX executable bin/test_package
/home/conan/w/cci_PR-3215/.conan/data/ignition-math/6.6.0/_/_/package/a8de599495731efa93328c25dca75843d6e89df0/lib/libignition-math6.a(Angle.cc.o): In function `_GLOBAL__sub_I_Angle.cc':
/home/conan/w/cci_PR-3215/.conan/data/ignition-math/6.6.0/_/_/build/a8de599495731efa93328c25dca75843d6e89df0/source_subfolder/src/Angle.cc:(.text.startup+0xc): undefined reference to `std::ios_base::Init::Init()'
/home/conan/w/cci_PR-3215/.conan/data/ignition-math/6.6.0/_/_/build/a8de599495731efa93328c25dca75843d6e89df0/source_subfolder/src/Angle.cc:(.text.startup+0x13): undefined reference to `std::ios_base::Init::~Init()'
```

The complete build line failing is:
```
/usr/bin/clang++ -lc++   -m64 -stdlib=libc++ -O3 -DNDEBUG        CMakeFiles/test_package.dir/test_package.cpp.o  -o bin/test_package   -L/home/conan/.conan/data/ignition-math/6.6.0/_/_/package/a8de599495731efa93328c25dca75843d6e89df0/lib  -Wl,-rpath,/home/conan/.conan/data/ignition-math/6.6.0/_/_/package/a8de599495731efa93328c25dca75843d6e89df0/lib /home/conan/.conan/data/ignition-math/6.6.0/_/_/package/a8de599495731efa93328c25dca75843d6e89df0/lib/libignition-math6.a -lm
```
which doesn't seem suspicious at all.
What I was able to do, which works but I don't know why, is replacing `-stdlib=libc++` with `-lc++`, resulting in:
```
/usr/bin/clang++ -lc++   -m64 -lc++ -O3 -DNDEBUG        CMakeFiles/test_package.dir/test_package.cpp.o  -o bin/test_package   -L/home/conan/.conan/data/ignition-math/6.6.0/_/_/package/a8de599495731efa93328c25dca75843d6e89df0/lib  -Wl,-rpath,/home/conan/.conan/data/ignition-math/6.6.0/_/_/package/a8de599495731efa93328c25dca75843d6e89df0/lib /home/conan/.conan/data/ignition-math/6.6.0/_/_/package/a8de599495731efa93328c25dca75843d6e89df0/lib/libignition-math6.a -lm
```
This last line builds correctly, and I can execute the test without problems.

I'm having a hard time understanding the difference between both and, in any case, how to fix it properly. Any ideas?

**Comment by joxoby** [comment]
@madebr you might have some insights here.

**Comment by madebr** [comment]
> @madebr you might have some insights here.

Can you reproduce it on your machine?

**Comment by joxoby** [comment]
> Can you reproduce it on your machine?

Yes, I'm using the docker image `conanio/clang50`

**Comment by joxoby** [comment]
The error that I'm getting:
```
/usr/bin/ld: /home/conan/.conan/data/ignition-math/6.6.0/_/_/package/3e4f7a83e79f6843d5f85d632d92d31bee9e4ef5/lib/libignition-math6.a(Angle.cc.o): in function `_GLOBAL__sub_I_Angle.cc':
Angle.cc:(.text.startup+0xc): undefined reference to `std::ios_base::Init::Init()'
/usr/bin/ld: Angle.cc:(.text.startup+0x13): undefined reference to `std::ios_base::Init::~Init()'
```
seems to be a bug in `libc++`: https://bugs.llvm.org/show_bug.cgi?id=28954.

So basically the problem is when using `libc++` with static libraries that use static initialization of basic streams.

What would be the best way to avoid a specific configuration, in this case, a static library with `clang` and `libc++` ?

**Comment by madebr** [comment]
I haven't looked into the clang problem yet, but please consider https://github.com/joxoby/conan-center-index/pull/1

**Comment by madebr** [comment]
> What would be the best way to avoid a specific configuration, in this case, static libraries with `clang` and `libc++` ?

**Comment by madebr** [comment]
> What would be the best way to avoid a specific configuration, in this case, static libraries with `clang` and `libc++` ?

That can be done in configure:
```python
def configure(self):
    if not self.options.shared and self.settings.compiler == "clang" and self.settings.compiler.libcxx = "libc++":
        raise ConanInvalidConfiguration("Due to a clang bug, this package cannot be used as a static library with libcxx==libc++.")
```

Eventually, you can filter finer by using `self.settings.compiler.version`.

**Comment by madebr** [comment]
I forgot to add the wrapper cmake script... oops

**Comment by madebr** [comment]
I cannot reproduce the `libc++` error when running the following command (from the recipe folder):
```
docker run -v $PWD:/recipe:z -ti conanio/clang50 /bin/sh -c "conan create /recipe ignition-math/6.6.0@ -s compiler.libcxx=libc++"
```

File: recipes/ignition-math/all/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 2.8.12)
+project(cmake_wrapper)
+
+include(conanbuildinfo.cmake)
+conan_basic_setup()
+
+set(CMAKE_MODULE_PATH ${CONAN_CMAKE_MODULE_PATH} ${CMAKE_SOURCE_DIR}/source_subfolder/build/cmake/modules ${CMAKE_MODULE_PATH})
+
+add_subdirectory(source_subfolder)
```
Comment (madebr) [inline recipes/ignition-math/all/CMakeLists.txt:9]:
```suggestion
cmake_minimum_required(VERSION 3.1)
project(cmake_wrapper)

include(conanbuildinfo.cmake)
conan_basic_setup()

add_subdirectory(source_subfolder)
```

**Comment by madebr** [comment]
Right now, it downloads both ignition-math and ignition-cmake.
The ignition-cmake is a collection of cmake scripts, would it make sense to create a `ignition-cmake` recipe and use it as a (build) requirement?
If there are other ignition package, then it would make sense to create a proper `ignition-cmake` recipe.

**Comment by joxoby** [comment]
Ok wow!

This is super interesting. What ended up fixing the problems was the CMake wrapper script. The wrapper added some linker flags to the compilation of the source code that were not passed in previously. For example, during the build of `Angle.cc`, the compiler is passing the `-stdlib=libc++` flag, which wasn't there before:
```
cd /home/conan/.conan/data/ignition-math/6.6.0/_/_/build/a3f2021885d1ab995129d41676fe7b8357771c06/source_subfolder/src && /usr/bin/clang++   -I/home/conan/.conan/data/pkgconf/1.7.3/_/_/package/961755854760682044d37dabfc2d998658edf961/include -I/home/conan/.conan/data/pkgconf/1.7.3/_/_/package/961755854760682044d37dabfc2d998658edf961/include/libpkgconf -I/home/conan/.conan/data/eigen/3.3.7/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/include -I/home/conan/.conan/data/eigen/3.3.7/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/include/eigen3 -I/home/conan/.conan/data/ignition-math/6.6.0/_/_/build/a3f2021885d1ab995129d41676fe7b8357771c06/source_subfolder/include -I/home/conan/.conan/data/ignition-math/6.6.0/_/_/build/a3f2021885d1ab995129d41676fe7b8357771c06/source_subfolder/core/include  -m64 -stdlib=libc++ -O3 -DNDEBUG   -Wall -Wextra -Wno-long-long -Wno-unused-value -Wfloat-equal -Wshadow -Winit-self -Wswitch-default -Wmissing-include-dirs -pedantic  -fPIC   -DPKGCONFIG_IS_STATIC -msse -msse2 -mfpmath=sse -msse3 -mssse3 -msse4.1 -msse4.2 -std=c++17 -o CMakeFiles/ignition-math6.dir/Angle.cc.o -c /home/conan/.conan/data/ignition-math/6.6.0/_/_/build/a3f2021885d1ab995129d41676fe7b8357771c06/source_subfolder/src/Angle.cc
```
Thanks @madebr !

**Comment by joxoby** [comment]
> Right now, it downloads both ignition-math and ignition-cmake.
The ignition-cmake is a collection of cmake scripts, would it make sense to create a ignition-cmake recipe and use it as a (build) requirement?
If there are other ignition package, then it would make sense to create a proper ignition-cmake recipe.

Yes, there will be more ignition packages coming, but I failed at making an `ignition-cmake` recipe when I tried. We can give it another shot.

**Comment by madebr** [comment]
Currently, `conanbuildinfo.cmake`, generated by the cmake generator acts as the cmake toolchain.
That's why it is important to add `cmake` to the generators.
If it is not added, the cmake defaults are used (`/usr/bin/cc;/usr/bin/c++;CFLAGS=;CXXFLAGS=;LDFLAGS`).

**Comment by madebr** [comment]
> > Right now, it downloads both ignition-math and ignition-cmake.
> > The ignition-cmake is a collection of cmake scripts, would it make sense to create a ignition-cmake recipe and use it as a (build) requirement?
> > If there are other ignition package, then it would make sense to create a proper ignition-cmake recipe.
> 
> Yes, there will be more ignition packages coming, but I failed at making an `ignition-cmake` recipe when I tried. We can give it another shot.

Please create one, the conan hooks will probably fail because it does not allow cmake scripts in the package.
But an exception can be added.

**Comment by joxoby** [comment]
> Please create one, the conan hooks will probably fail because it does not allow cmake scripts in the package.
But an exception can be added.

Ok, working on it now.

**Comment by joxoby** [comment]
Draft PR for `ignition-cmake`: https://github.com/conan-io/conan-center-index/pull/3268

**Comment by madebr** [comment]
See https://github.com/joxoby/conan-center-index/pull/2,
it will fail because `ignition-cmake` is not available on cci yet

**Pull request closed without merging**